File: julia/src/benchmarks/Knuckeotide.jl
mutable struct Knuckeotide <: AbstractBenchmark
    seq::String
    result::String

    function Knuckeotide()
        new("", "")
    end
end

name(b::Knuckeotide)::String = "Knuckeotide"

function frequency(seq::String, len::Int)::Tuple{Int, Dict{String, Int}}
    n = length(seq) - len + 1
    table = Dict{String, Int}()

    for i in 1:n
        sub = seq[i:i+len-1]
        table[sub] = Base.get(table, sub, 0) + 1  
    end

    return (n, table)
end

function find_seq(b::Knuckeotide, seq::String, s::String)
    len_s = length(s)
    n, table = frequency(seq, len_s)
    s_lower = lowercase(s)
    count = Base.get(table, s_lower, 0)  

    s_upper = uppercase(s)
    b.result *= "$count\t$s_upper\n"
end

function sort_by_freq(b::Knuckeotide, seq::String, len::Int)  
    n, table = frequency(seq, len)

    pairs = collect(table)
    sort!(pairs, by = x -> (-x[2], x[1]))

    for (key, value) in pairs
        percent = (value * 100.0) / n
        key_upper = uppercase(key)
        b.result *= "$key_upper $(@sprintf("%.3f", percent))\n"
    end
    b.result *= "\n"
end

function prepare(b::Knuckeotide)

    fasta = Fasta()
    fasta.n = Helper.config_i64("Knuckeotide", "n")

    run(fasta, Int64(0))

    fasta_result = String(take!(fasta.io))
    lines = split(fasta_result, '\n')
    three = false
    seq_parts = String[]

    for line in lines
        if startswith(line, ">THREE")
            three = true
            continue
        end
        if three && !startswith(line, ">") && !isempty(line)
            push!(seq_parts, line)
        end
    end

    b.seq = join(seq_parts)
end

function run(b::Knuckeotide, iteration_id::Int64)
    for i in 1:2
        sort_by_freq(b, b.seq, i)
    end

    searches = ["ggt", "ggta", "ggtatt", "ggtattttaatt", "ggtattttaatttatagt"]
    for s in searches
        find_seq(b, b.seq, s)
    end
end

checksum(b::Knuckeotide)::UInt32 = Helper.checksum(b.result)

File: nim/src/benchmarks/knuckeotide.nim
import std/[strutils, sequtils, tables, algorithm, strformat, math]
import ../benchmark
import ../helper
import fasta

type
  Knuckeotide* = ref object of Benchmark
    seq: string
    resultStr: string

proc newKnuckeotide(): Benchmark =
  Knuckeotide()

method name(self: Knuckeotide): string = "CLBG::Knuckeotide"

method prepare(self: Knuckeotide) =
  var fastaInst = cast[Fasta](newFasta())

  fastaInst.n = self.config_val("n")
  fastaInst.prepare()
  fastaInst.run(0)

  let res = fastaInst.get_result()

  var three = false
  self.seq = ""

  for line in res.splitLines():
    if line.startsWith(">THREE"):
      three = true
      continue
    if three:
      self.seq.add(line)

  self.resultStr = ""

proc frequency(seq: string, length: int): (int, Table[string, int]) =
  let n = seq.len - length + 1
  var table = initTable[string, int]()

  for i in 0..<n:
    let sub = seq[i..<i+length]
    table.mgetOrPut(sub, 0).inc()

  result = (n, table)

method sortByFreq(self: Knuckeotide, seq: string, length: int) =
  let (n, table) = frequency(seq, length)

  var pairs: seq[(string, int)]
  for key, value in table.pairs():
    pairs.add((key, value))

  pairs.sort do (a, b: (string, int)) -> int:
    if a[1] == b[1]:
      return cmp(a[0], b[0])
    return cmp(b[1], a[1])

  for (key, value) in pairs:
    let percent = (value.float * 100.0) / n.float
    self.resultStr.add(key.toUpperAscii() & " " & formatFloat(percent,
        ffDecimal, 3) & "\n")

  self.resultStr.add("\n")

method findSeq(self: Knuckeotide, seq: string, s: string) =
  let length = s.len
  let (n, table) = frequency(seq, length)
  let sLower = s.toLowerAscii()
  let count = table.getOrDefault(sLower, 0)

  self.resultStr.add($count & "\t" & s.toUpperAscii() & "\n")

method run(self: Knuckeotide, iteration_id: int) =
  for i in 1..2:
    self.sortByFreq(self.seq, i)

  let searches = @["ggt", "ggta", "ggtatt", "ggtattttaatt", "ggtattttaatttatagt"]
  for s in searches:
    self.findSeq(self.seq, s)

method checksum(self: Knuckeotide): uint32 =
  checksum(self.resultStr)

registerBenchmark("CLBG::Knuckeotide", newKnuckeotide)
